libraries/AP_IBus_Telem/AP_IBus_Telem.cpp

```diff
@@ -0,0 +1,674 @@
+        break;
+
+    case IBUS_SENSOR_TYPE_FUEL:
+        value.uint16 = get_battery_or_fuel_level_pct();
+        break;
+
+    case IBUS_SENSOR_TYPE_RPM:
+    case IBUS_SENSOR_TYPE_RPM_FLYSKY:
+        value.uint16 = get_rpm();
+        break;
+
+    case IBUS_SENSOR_TYPE_COMPASS_HEADING:
+        value.uint16 = AP::ahrs().yaw_sensor * 0.01;
+        break;
+
+    case IBUS_SENSOR_TYPE_CLIMB_RATE:
+    case IBUS_SENSOR_TYPE_VERTICAL_SPEED:
+        value.int16 = AP::baro().get_climb_rate() * 100;
+        break;
+
+    case IBUS_SENSOR_TYPE_COG:
+        value.uint16 = AP::gps().ground_course() * 100;
+        break;
+
+    case IBUS_SENSOR_TYPE_GPS_STATUS:
+        value.byte[0] = get_gps_status();
+        value.byte[1] = AP::gps().num_sats();
+        break;
+
+    case IBUS_SENSOR_TYPE_ACC_X:
+        value.int16 = get_unbiased_acceleration().x * 1000;
+        break;
+
+    case IBUS_SENSOR_TYPE_ACC_Y:
+        value.int16 = get_unbiased_acceleration().y * 1000;
+        break;
+
+    case IBUS_SENSOR_TYPE_ACC_Z:
+        value.int16 = get_unbiased_acceleration().z * 1000;
+        break;
+
+    case IBUS_SENSOR_TYPE_ROLL:
+        value.int16 = AP::ahrs().roll_sensor;
+        break;
+
+    case IBUS_SENSOR_TYPE_PITCH:
+        value.int16 = AP::ahrs().pitch_sensor;
+        break;
+
+    case IBUS_SENSOR_TYPE_YAW:
+        value.int16 = AP::ahrs().yaw_sensor;
+        break;
+
+    case IBUS_SENSOR_TYPE_GROUND_SPEED:
+        value.uint16 = AP::gps().ground_speed_cm();
+        break;
+
+    case IBUS_SENSOR_TYPE_GPS_DIST:
+        value.uint16 = get_distance_from_home_m();
+        break;
+
+    case IBUS_SENSOR_TYPE_ARMED:
+        value.uint16 = AP::arming().is_armed();
+        break;
+
+    case IBUS_SENSOR_TYPE_FLIGHT_MODE:
+        value.uint16 = get_vehicle_mode();
+        break;
+
+    case IBUS_SENSOR_TYPE_SPEED:
+        value.uint16 = AP::airspeed()->get_airspeed();
+        break;
+
+    case IBUS_SENSOR_TYPE_TEMPERATURE_PRESSURE: {
+        const uint32_t pressure = AP::baro().get_pressure();
+        const uint32_t temperature = (AP::baro().get_temperature() + 40) * 10;
+        value.uint32 = pressure | (temperature << 19);
+        break;
+    }
+
+    case IBUS_SENSOR_TYPE_GPS_LAT:
+        value.int32 = AP::gps().location().lat;
+        break;
+
+    case IBUS_SENSOR_TYPE_GPS_LNG:
+        value.int32 = AP::gps().location().lng;
+        break;
+
+    case IBUS_SENSOR_TYPE_GPS_ALT:
+        value.int32 = AP::gps().location().alt;
+        break;
+
+    case IBUS_SENSOR_TYPE_ALT:
+        value.int32 = AP::baro().get_altitude() * 100;
+        break;
+
+    default:
+        value.int32 = 0;
+    }
+
+    return value;
+}
+
+uint16_t AP_IBus_Telem::get_average_cell_voltage_cV()
+{
+    if (!AP::battery().has_cell_voltages()) {
+        return 0;
+    }
+
+    const auto &cell_voltages = AP::battery().get_cell_voltages();
+    const uint8_t number_of_cells = ARRAY_SIZE(cell_voltages.cells);
```

Comment:
defensive programming says we should check for zero to avoid the divide-by-zero case below.